### ISSUE TITLE
Logging to file

### ISSUE THREAD

**Issue by RafalBuchner** [googlefonts/fontmake#647]
Hi,
Is it possible to save logging's output directly to a txt file?

When I'm using this:

`fontmake -o fontfolder -m my.designspace > myVariableReport.txt` 

it creates an empty txt file.

I've also been trying to create a script with FontProject class within RoboFont, where I'm overwriting FontProject's logging by adding the `logging.StreamHandler()` with a path to some txt file. Unfortunately, only one line is being recorded, "Building variable font /path/to/this/font.ttf" nothing else. 
It would be super cool to be able to save those reports with fontmake, since it gives a good overview during the test exports and helps to communicate with the rest of the design team.

Best,
R

**Comment by punchcutter** [comment]
Sometimes you want `2>&1` to redirect stderr to stdout, but here you need `2>` like

`fontmake -o fontfolder -m my.designspace 2> myVariableReport.txt`

**Comment by anthrotype** [comment]
one of the reason i don't like fontmake to be used as a library is that initializing a FontProject class calls logging.basicConfig (which mutates global state). I'd like to fix that someday.

What @punchcutter says is correct; logging module by default prints to the stderr stream (and fontmake simply calls logging.basicConfig), so from the console you want to redirect stderr to a file with `2> fontmake-log.txt`

When using the FontProject class from python code, I believe you should be able to configure logging _before_ you instantiate the first FontProject; that is, attach a handler to the root logger and set the desired verbosity.
The basicConfig only configures the root logger once if it hasn't been configured yet, and should do nothing if it already has handlers configured.

What does your logging setup code look like?

**Comment by madig** [comment]
Have you considered using ufo2ft's functions directly instead of messing with fontmake?

**Comment by RafalBuchner** [comment]
@anthrotype this is my code, first time with logging, maybe I messed something up

```fp = FontProject(verbose="DEBUG", validate_ufo=True)
		my_logger = logging.getLogger(fontmake.font_project.__name__)
		formatter = logging.Formatter("%(asctime)s - %(name)s - %(levelname)s - %(message)s")
		handler = logging.StreamHandler(stream)
		handler.setFormatter(formatter)
		handler.setLevel(logging.DEBUG)
		my_logger.addHandler(
				handler
			)
		fp.build_variable_font(
            self.ds_path,
            output_path=None,
            output_dir=self.output_path,
            ttf=True,
            optimize_gvar=True,
            use_production_names=None,
            reverse_direction=True,
            conversion_error=None,
            feature_writers=None,
            cff_round_tolerance=None,
            debug_feature_file=None,
            )
```

**Comment by RafalBuchner** [comment]
@madig not really, but still, it is requested from my client to generate the report along with the variable font

**Comment by anthrotype** [comment]
you are only adding a handler to the fontmake logger, that's why you only get the messages from the latter, but not also from other libraries that emit logging messages (fonttools, cu2qu, ufo2ft, etc.)
You want to configure the root logger instead (what you get from a bare `logging.getLogger()` call).
Do that before you instantiate a new FontProject class, as the latter in turn calls `logging.basicConfig` and would change the root logger config (as I explained above).
